### PR TITLE
GMG: coarse smoother and linear operator

### PR DESCRIPTION
**Pull request by davydden**
extend GMG clases to (i) do smoothing at the coarse level; (ii) be used as LinearOperator;

The latter is a bit hacky, but i don't see any cleaner way for now.

### CODE REVIEW

**Comment by tjhei** [comment]
I think it would be better to implement a new class derived from ``MGCoarseGridBase`` that takes a smoother as a template parameter (similar to ``MGCoarseGridLACIteration`` but that doesn't invoke a solver).

**Comment by kronbichler** [comment]
I agree with @tjhei that we should try to keep the interface clean and rather provide a new class derived from `MGCoarseGridBase` that calls `Smoother::smooth`. On the other hand, given that writing such a class on a case to case basis is so easy, just using 5-10 lines of code, I'm not really convinced that we should spend too much effort on providing too many different cases rather than teaching users how to implement them on their own. I must have ten own implementations of a coarse iterative solvers in my various projects rather than `MGCoarseGridLACIteration` because it is so simple to write those 20 lines that I never bothered to remember the interface specified by `MGCoarseGridLACIteration`. If we add two or three more classes, implementing the interface on your own rather than finding the correct template will be even more tempting...

**Comment by davydden** [comment]
ok, I will rework into a new class for smoothing derived from `MGCoarseGridBase`.

**Comment by davydden** [comment]
@tjhei @kronbichler i updated the PR.

File: include/deal.II/multigrid/mg_coarse.h [hunk 1]
```diff
@@ -28,6 +28,47 @@ DEAL_II_NAMESPACE_OPEN
...
 
 /**
+ * Coarse grid solver using preconditioners only. This is a little wrapper,
+ * transforming a preconditioner into
+ * a coarse grid solver.
+ *
+ * @author Denis Davydov, 2016.
+ */
+template<class VectorType = Vector<double> >
+class MGCoarseGridPreconditioner :  public MGCoarseGridBase<VectorType>
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_coarse.h]: Should we really call this method `Preconditioner`? I would find it more intuitive to call this method `MGCoarseGridFromSmoother` or similar to indicate that we take the `smooth` operation. When thinking about a preconditioner, my view would be that this class needs to be templated on an arbitrary `PreconditionerType` type that gets fed to the initialize class. Other opinions?

Comment (tjhei) [inline include/deal.II/multigrid/mg_coarse.h]: ``MGCoarseGridApplySmoother``?

Comment (davydden) [inline include/deal.II/multigrid/mg_coarse.h]: `MGCoarseGridApplySmoother` sounds good to me, will rename...

**Comment by davydden** [comment]
i renamed the class.

File: doc/news/changes.h [hunk 2]
```diff
@@ -428,6 +428,13 @@ SparsityPattern.
 <h3>Specific improvements</h3>
 
 <ol>
+ <li> New: PreconditionMG can now be used as a LinearOperator. Implement
+ MGCoarseGridApplySmoother class to do a few steps of a smoother at
+ the coarsest level.
```
Comment (kronbichler) [inline doc/news/changes.h]: I suggest you split this into two list items since the coarse grid smoother and LinearOperator parts are unrelated.

**Comment by kronbichler** [comment]
I like the new name. Apart from my minor comment, please run the indentation script. Then this should be good to go.

**Comment by davydden** [comment]
i fixed indentation and the entry in changes.h

File: include/deal.II/multigrid/mg_coarse.h [hunk 3]
```diff
@@ -28,6 +28,47 @@ DEAL_II_NAMESPACE_OPEN
 /*@{*/
 
 /**
+ * Coarse grid solver using preconditioners only. This is a little wrapper,
+ * transforming a preconditioner into
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_coarse.h]: preconditioner -> smoother

File: include/deal.II/multigrid/mg_coarse.h [hunk 4]
```diff
@@ -28,6 +28,47 @@ DEAL_II_NAMESPACE_OPEN
...
+class MGCoarseGridApplySmoother :  public MGCoarseGridBase<VectorType>
+{
+public:
+  /**
+   * Default constructor.
+   */
+  MGCoarseGridApplySmoother ();
+
+  /**
+   * Constructor. Store preconditioner for later use.
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_coarse.h]: Say "Store a pointer to the smoother for later use", to make clear that this object does not actually store the preconditioner and rather must be kept alive.

File: include/deal.II/multigrid/mg_coarse.h [hunk 5]
```diff
@@ -28,6 +28,47 @@ DEAL_II_NAMESPACE_OPEN
...
+  /**
+   * Implementation of the abstract function.
+   */
+  void operator() (const unsigned int level,
+                   VectorType         &dst,
+                   const VectorType   &src) const;
+
+private:
+  /**
+   * Reference to the preconditioner.
```
Comment (tjhei) [inline include/deal.II/multigrid/mg_coarse.h]: smoother

**Comment by tjhei** [comment]
Wouldn't it also make sense to apply a smoother more than once?

**Comment by davydden** [comment]
@tjhei you already can by
```
    mg_coase_grid_smoother.set_steps(2);
```

File: tests/multigrid/step-16-50-mpi-smoother.cc [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the deal.II distribution.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+// Same as step-16-50, but use Jacobi smoother at the coarsest grid via MGCoarseGridApplySmoother.
+// In this particular case, the number of iterations untill convergence is
```
Comment (tjhei) [inline tests/multigrid/step-16-50-mpi-smoother.cc]: typo "until"

File: tests/multigrid/step-16-50-mpi-linear-operator.with_mpi=true.with_trilinos=true.with_p4est=true.mpirun=3.output [hunk 7]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (tjhei) [inline tests/multigrid/step-16-50-mpi-linear-operator.with_mpi=true.with_trilinos=true.with_p4est=true.mpirun=3.output]: the output file needs ``with_cxx11=on.``, right?

**Pull request merged**